x-pack/plugins/endpoint/server/services/endpoint/endpoint_query_builders.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { KibanaRequest } from 'kibana/server';
+import { EndpointAppContext } from '../../types';
+
+export class AllEndpointsQueryBuilder {
+  private readonly request: KibanaRequest<any, any, any>;
+  private readonly endpointAppContext: EndpointAppContext;
+
+  constructor(request: KibanaRequest<any, any, any>, endpointAppContext: EndpointAppContext) {
+    this.request = request;
+    this.endpointAppContext = endpointAppContext;
+  }
+
+  async toQueryParams(): Promise<Record<string, unknown>> {
+    const paging = await this.paging();
+    return {
+      body: {
+        query: this.queryBody(),
+        collapse: {
+          field: 'machine_id',
+          inner_hits: {
+            name: 'most_recent',
+            size: 1,
+            sort: [{ created_at: 'desc' }],
+          },
+        },
+        aggs: {
+          total: {
+            cardinality: {
+              field: 'machine_id',
+            },
+          },
+        },
+        sort: [
+          {
+            created_at: {
+              order: 'desc',
+            },
+          },
+        ],
+      },
+      from: paging.pageIndex * paging.pageSize,
+      size: paging.pageSize,
+      index: 'endpoint-agent*',
```

Comment:
Should this be configurable somehow?

Reply:
> Should this be configurable somehow?

I think we should make it conventional for now and agree what it will be called, if we make it configurable a lot of system will have to change every time some changes this value. fleet, endpoint generator etc. I can make it a constant.